경북대 BE_박신영 2단계 - 관리자 화면

parknew0 opened the pull request:
이전 피드백 반영하고, 2단계 관리자 화면까지 구현해봤습니다!

Thymeleaf랑 HTML 기반으로 상품 등록/수정/삭제/목록 조회 기능 넣었습니다.
시간에 쫓기듯 만들어서 기능 위주로 빠르게 만들어서 이번 구현은 저도 많이 아쉬움이 많이 남는데요,

시간 여유 생기면 AJAX로 리팩토링도 해보고 싶습니다..

parknew0 committed:
docs(readme): add feature list for step 1

parknew0 committed:
docs(readme): update feature list for step 1

parknew0 committed:
feat(product): implement GET /api/products to return product list

parknew0 committed:
feat(product): implement GET /api/products/{id} for single product search

parknew0 committed:
feat(product): implement POST /api/products for product creation

parknew0 committed:
feat(product): implement PUT /api/products/{id}

parknew0 committed:
feat(product): implement DELETE /api/products/{id}

parknew0 committed:
style: reformat code to use 4-space indentation

parknew0 committed:
style(controller): ResponseEntity 응답 형식 HttpStatus 기반으로 통일

parknew0 committed:
style: 모든 파일 끝에 누락된 개행 문자 추가

parknew0 committed:
refactor(repository): 초기 더미 데이터 제거 및 sequence 초기값 1로 수정

parknew0 committed:
style(repository): 불필요한 줄바꿈 제거하여 코드 정리

parknew0 committed:
refactor(service): getProductById를 findProductById로 메서드 네이밍 명확화

parknew0 committed:
refactor(product): 생성자에서 id 제거하고 repository에서 setId로 주입하도록 변경

parknew0 committed:
refactor: update 요청 시 상품이 없을 경우 예외 처리하도록 변경

parknew0 committed:
refactor(service): findProductById에서 null 대신 NoSuchElementException 던지도록 수정

parknew0 committed:
style(controller): 개행 오타 수정

parknew0 committed:
docs(readme): add feature list for step 2

parknew0 committed:
feat(admin): 상품 등록 폼 및 등록 기능 구현

parknew0 committed:
feat(admin): 상품 수정 페이지 및 수정 기능 추가

parknew0 committed:
feat(admin): 삭제 기능 추가 및 수정 버튼 수정

Rok93 commented on README.md:
```diff
@@ -1,9 +1,18 @@
-## 구현 기능 목록 (1단계 - 상품 API)
+## 구현할 기능 목록 (1단계 - 상품 API)
 
 - [ ] 상품 목록 조회 API (GET /api/products)
 - [ ] 상품 단건 조회 API (GET /api/products/{productId})
 - [ ] 상품 등록 API (POST /api/products)
 - [ ] 상품 수정 API (PUT /api/products/{productId})
 - [ ] 상품 삭제 API (DELETE /api/products/{productId})
+- [ ] JSON 요청 및 응답 포맷 설정
+- [ ] 메모리 저장소(Map<Long, Product>) 구현
+
+## 구현할 기능 목록 (2단계 - 관리자 화면)
+
+- [ ] 관리자 상품 목록 조회 화면
+- [ ] 상품 등록 화면 및 등록 기능
+- [ ] 상품 수정 화면 및 수정 기능
+- [ ] 상품 삭제 기능
```
기능 구현 목록표 잘 작성해주셨는데요.  👍👍 
가능한 커밋단위도 이 기능 구현 목록표에 맞춰서 하나씩 같이 체크를 해나가시면 좋을 것 같아요. 😃

그렇다면 조금 더 살아있는 문서가 될 수 있을 것 같네요. 😉

Rok93 commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -5,57 +5,55 @@
+        Long id = sequence++;
+        Product newProduct = new Product(product.getName(), product.getPrice(),
+                product.getImageUrl());
+        newProduct.setId(id);
```
[프로그래밍 요구사항](https://edu.nextstep.camp/s/UH6H8Iba/ls/Mvfn9N7o)의 `객체지향 생활체조 원칙`에 따르면 


[1단계 코멘트](https://github.com/next-step/spring-gift-product/pull/74#discussion_r2166528962)를 다시 한번 확인해주시면 감사하겠습니다. 😃 

코멘트 확인하시면 간단하게라도 이모지 남겨주시면 제가 잘 확인하고 지나가셨는지 확인하기 쉬울 것 같아요. 
가능하면 확인하신 코멘트에 대해서는 간단히 👍 혹은 👀 이모지 등을 남겨주시면 좋을 것 같아요. 🙏

parknew0 replied:
규칙 9: getter/setter/프로퍼티를 쓰지 않는다. 가 위배된 부분인것같네요! 해당 부분 다시 공부해서 익히겠습니다

Rok93 commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -5,57 +5,55 @@
-    if (!products.containsKey(id)) {
-      return false;
+        Product updated = new Product(name, price, imageUrl);
+        updated.setId(id);
```
업데이트를 새로운 객체를 만들어서 대체하실꺼라면 
주생성자를 통해서 id까지 할당을 받는게 조금 더 깔끔하지 않을까요? 😃


```suggestion
        Product updated = new Product(id, name, price, imageUrl);
```

parknew0 replied:
지당하신 말씀입니다. 이전 코드는 일을 두번 나눠했네요..!

Rok93 commented on src/main/resources/templates/admin/product/list.html:
```diff
@@ -0,0 +1,40 @@
+<!DOCTYPE html>
+<html xmlns:th="http://www.thymeleaf.org">
```
템플릿 구현 잘 해주셨어요! 👍

<img width="332" alt="image" src="https://github.com/user-attachments/assets/cbd60140-8286-4411-b38c-c0b47f66293e" />

다만 조금 아쉬운건 css 요소를 지정한게 없다보니 조금 테이블이 보기 어렵네요. 🥲
적절히 css 요소를 추가해서 테이블만 조금 더 가독성을 높여주면 좋을 것 같아요. 😃

parknew0 replied:
수정했습니다!

Rok93 commented on src/main/java/gift/controller/admin/AdminProductController.java:
```diff
@@ -0,0 +1,68 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+@RequestMapping("/admin/products")
```
ProductController에서는 `@RestController`를 AdminProductController에서는 `@Controller`를 활용해주셨는데 
이 둘의 차이점은 무엇인가요? 😃

추가적으로 이 두 애너테이션의 내부 코드를 확인해보시면 좋을 것 같아요. 😉

parknew0 replied:
@Controller는 html같은 뷰 반환용이고, 
@RestController는 JSON 같은 본문 반환용입니다. 확인해보겠습니다 :))

Rok93 replied:
확인해보시고 어떤 차이점이 있는지 가볍게라도 코멘트로 남겨주시면 좋을 것 같아요. 😃 (사소해보이지만 이런 포인트를 알고 지나가냐 모르고 지나가냐가 누적되면 꽤 큰 차이? 가 나는 것 같아요. )

마지막 리뷰 요청 마감일이 오늘이다보니 일단 3단계 미션을 먼저 진행해주시고 미션 진행 후 시간이되면 이 부분도 더 알아보시면 좋을 것 같아요.

Rok93 commented on src/main/java/gift/controller/admin/AdminProductController.java:
```diff
@@ -0,0 +1,68 @@
+    public String showProductList(Model model) {
+        List<ProductResponse> products = productService.getAllProducts();
+        model.addAttribute("products", products);
+        return "admin/product/list";
```
추가적으로 이번 미션에서는 `서버사이드 렌더링`을 하셨는데, 그렇다면 `클라이언트 사이드 렌더링`은 무엇인지에 대해서도 가볍게 살펴보시면 좋을 것 같아요. 😃

parknew0 replied:
클라이언트사이드 렌더링은 json같은 데이터만 받아서 화면구성은 브라우저에서 처리하는 방식입니다!
이전에 리액트와 블레이저를 이용해 프론트엔드 구현을 해본 적이 있어서 조금 알고있습니다!

Rok93 commented on src/main/java/gift/dto/ProductRequest.java:
```diff
@@ -2,22 +2,43 @@
 
 public class ProductRequest {
```
자바 14부터 [record class](https://colevelup.tistory.com/28)라는 것이 생겼는데요. 
지금과 같은 DTO 클래스는 record 클래스로 대체해볼 수 있을 것 같아요. 

record class가 어떤 특징을 가지는지 학습해보고 이해하신 내용을 가볍게 코멘트로 남겨주시면 좋을 것 같아요. 😁

parknew0 replied:
레코드 클래스를 사용하는 이유로는, 불변성 보장, 간결성, 명확한 의도라고 하는데요.
가장 중요한건 불변성 보장을 통해서 안정성을 높이는 것 같네요.

지금 setter를 활용하고 있는 현재의 코드는 객체 생성 후에 내부 데이터가 변경될 수 있어 위험하다고 이해했습니다.
record class로 변경한 뒤에 기존 코드를 사용하던 다른 코드들도 전부 수정을 진행하겠습니다!

Rok93 commented on src/main/java/gift/controller/admin/AdminProductController.java:
```diff
@@ -0,0 +1,68 @@
+    @GetMapping("/{id}/edit")
+    public String showEditForm(@PathVariable Long id, Model model) {
+        ProductResponse product = productService.findProductById(id);
+        ProductRequest request = new ProductRequest();
+        request.setId(product.getId());
+        request.setName(product.getName());
+        request.setPrice(product.getPrice());
+        request.setImageUrl(product.getImageUrl());
```
[프로그래밍 요구사항](https://edu.nextstep.camp/s/UH6H8Iba/ls/Mvfn9N7o)을 지키며 미션을 수행해보시면 좋을 것 같아요. 😃
일단 setter 메서드를 없애고 생성자를 통해 Request 객체를 만들어 보면 어떨까요?

Rok93 commented on src/main/java/gift/controller/admin/AdminProductController.java:
```diff
@@ -0,0 +1,68 @@
+        request.setPrice(product.getPrice());
+        request.setImageUrl(product.getImageUrl());
+
+        model.addAttribute("product", request);
```
생각해보니 이 부분은 굳이 **ProductResponse** 를 `ProductRequest`로 변경하여 ModelAttribute로 넘겨줄 필요가 없을 것 같은데요? 🤔

조회해온 `ProductResponse` 를 그대로 product Attribute로 전달하면 되지 않을까요? 

```suggestion
        model.addAttribute("product", product);
```

parknew0 replied:
edit 페이지라서 수정된 데이터 내용을 클라이언트에서 서버로 전달해주는 것만 생각했었는데
생각해보니 서버에서 클라이언트로 edit 페이지 폼을 보내주는 역할이었네요!
제안해주신 대로 수정했습니다. 감사합니다!

Rok93 commented on src/main/resources/templates/admin/product/form.html:
```diff
@@ -0,0 +1,49 @@
+
+<!-- 수정 폼: ID가 있는 경우 (기존 상품 수정) -->
+<form th:if="${product.id != null}"
+      th:action="@{'/admin/products/' + ${product.id}}"
```
`product.id`가 null인지 아닌지 여부로 분기처리를 하고있으신 것 같은데요. 

상품 생성 폼과 상품 수정 폼은 분리하면 어떨까요? 🤔
분리하게된다면 생성 폼의 경우에는 `product.id`가 null이고   수정 폼의 경우에는 `product.id`가 null이 아니기 때문에 view에서 이런 분기처리를 없앨 수 있을 것 같아요. 😃

parknew0 replied:
두 페이지로 분리했습니다!

Rok93 replied:
현재는 이 두 폼이 거의 똑같지만 추후 이 두 폼은 각각 독립적으로 성장할 확률이 있다고 생각해요.
그렇기 때문에 이전과 같은 View는 각각의 케이스가 발전해감에따라 분기처리가 더 늘어날 가능성이 높다고 생각해요.  😃

Rok93 commented on src/main/resources/templates/admin/product/list.html:
```diff
@@ -0,0 +1,40 @@
+  </tr>
+  </thead>
+  <tbody>
+  <tr th:each="product : ${products}">
```
> 시간 여유 생기면 AJAX로 리팩토링도 해보고 싶습니다..

시간이 되실지모르겠지만, 말씀하신김에 도전해보면 어떨까요? 😁

parknew0 replied:
AJAX는 시간상 이번 도전은 힘들 것 같습니다 ㅜㅜ

Rok93 replied:
엄청 중요한 부분은 아니라고 생각하기 때문에 꼭 다뤄야할 부분은 아니라고 생각해요. 
아쉽지만 시간관계상 ajax는 적용하지 않는걸로 이해하고있을게요!

Rok93 reviewed (changes requested):
신영님 2단계도 빠르게 잘 구현해주셨네요. 👍👍
몇몇 코멘트 남겨두었는데 확인 부탁드릴게요. 😃

그리고 AJAX로 전환하는 작업도 도전해보실지 말지 고민해보시고 시간이 괜찮다면 진행해보셔도 좋을 것 같아요. 😉

parknew0 committed:
refactor: setter 없에고 assignId 를 도입, 객체지향생활체조 원칙 9번을 준수하려는 시도

parknew0 committed:
refactor(dto): ProductRequest를 record로 전환하고 관련 코드 수정
- ProductRequest: 기존 class에서 record로 전환
- AdminProductController: 기본 생성자 및 setter 사용 코드를 record 생성자 방식으로 수정
- ProductService: record 접근자(getter) 형식에 맞춰 메서드 호출 방식 변경 (e.g., getName() -> name())

parknew0 committed:
refactor(admin): 수정 폼 조회 시 불필요한 DTO 변환 제거

parknew0 committed:
refactor(admin): 관리자 페이지 UI/UX 전반적 개선

아래와 같은 리팩토링 및 스타일링을 진행했습니다.
- 상품 등록/수정 폼을 별도 파일로 분리하여 뷰 로직 단순화
- 상품 목록 테이블의 가독성을 높이기 위한 CSS 스타일 추가
- 폼 내 버튼들의 시각적 불일치 문제 해결

Rok93 reviewed (approved):
신영님 피드백 잘 반영해주셨네요. 👍
2단계도 충분히 잘 구현해주신 것 같아서 이만 머지하도록할게요. 😉
이어서 3단계 진행해주세요. 🚀🚀
